src/test/java/org/apache/maven/buildcache/its/Issue87Test.java

```diff
@@ -89,7 +89,7 @@ void simple(Verifier verifier) throws VerificationException, IOException {
                 1,
                 projectsInputInfo.getItems().stream()
                         .filter(item -> "pluginDependency".equals(item.getType()))
-                        .filter(item -> ("org.apache.maven.plugins:maven-dependency-plugin:maven-plugin|"
+                        .filter(item -> ("org.apache.maven.plugins:maven-dependency-plugin:maven-plugin|0|"
```

Comment:
It will be nice to add a test for two plugin declarations in the build. Something like:
* module 1 -> produces dependency 1
* module 2 -> produces dependency 2
* module 3 -> declares two resource plugins (different executions). The first is with the dependency on module 1, the second is with module 2.

Expected: changing any plugin yields a new checksum, and the project rebuilds. The previous implementation should fail such a test because the checksum contribution by the first plugin could be overridden by the second, unchanged plugin.

Reply:
Hi
I'm not sure why it's throwing an exception. Could it be the enforcer or another similar plugin? What version of Maven is that? Please paste the full stack trace. I'm pretty confident that Maven allows duplicated declarations. This is my local project (3.9.5) and it runs just fine.
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>TestPom</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>11</maven.compiler.source>
        <maven.compiler.target>11</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <executions>
                    <execution>
                        <id>exec1</id>
                        <goals>
                            <goal>analyze</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <executions>
                    <execution>
                        <id>exec2</id>
                        <goals>
                            <goal>analyze-only</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```

Reply:
@AlexanderAshitkin , just apply this patch to the last commit of this PR to reproduce:

```patch
Subject: [PATCH] Simplify `keyPrefixOccurrenceIndex`
---
Index: src/test/projects/mbuildcache-87/top/pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/projects/mbuildcache-87/top/pom.xml b/src/test/projects/mbuildcache-87/top/pom.xml
--- a/src/test/projects/mbuildcache-87/top/pom.xml	(revision e50d052a242d195f28d553fc8216b2d6035a724d)
+++ b/src/test/projects/mbuildcache-87/top/pom.xml	(date 1718987310487)
@@ -42,6 +42,7 @@
     <modules>
         <module>module1</module>
         <module>module2</module>
+        <module>module3</module>
     </modules>
 
 </project>
Index: src/test/projects/mbuildcache-87/top/module3/pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/projects/mbuildcache-87/top/module3/pom.xml b/src/test/projects/mbuildcache-87/top/module3/pom.xml
new file mode 100644
--- /dev/null	(date 1718987535540)
+++ b/src/test/projects/mbuildcache-87/top/module3/pom.xml	(date 1718987535540)
@@ -0,0 +1,58 @@
+<!--
+
+    Copyright 2021 the original author or authors.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+         http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>module3</artifactId>
+    <packaging>jar</packaging>
+
+    <parent>
+        <groupId>org.apache.maven.caching.test.mbuildcache-87</groupId>
+        <artifactId>top</artifactId>
+        <version>0.0.1-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-dependency-plugin</artifactId>
+                <dependencies>
+                    <dependency>
+                        <groupId>${project.groupId}</groupId>
+                        <artifactId>module2</artifactId>
+                        <version>${project.version}</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-dependency-plugin</artifactId>
+                <dependencies>
+                    <dependency>
+                        <groupId>${project.groupId}</groupId>
+                        <artifactId>module3</artifactId>
+                        <version>${project.version}</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
+        </plugins>
+    </build>
+</project>

```

Reply:
Here is the full stacktrace:

```
[INFO] Error stacktraces are turned on.
[DEBUG] Reading global settings from 'settings.xml'
[DEBUG] Reading project settings from '/home/rhousni/projects/maven-build-cache-extension/target/mvn-cache-tests/Issue87Test/simple/project/top/.mvn/settings.xml'
[DEBUG] Reading user settings from '/home/rhousni/.m2/settings.xml'
[DEBUG] Using manager EnhancedLocalRepositoryManager with priority 10.0 for /home/rhousni/projects/maven-build-cache-extension/target/local-repo
[INFO] Scanning for projects...
[ERROR] Some problems were encountered while processing the POMs
[ERROR] The build could not read 1 project -> [Help 1]
org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs
    at org.apache.maven.project.DefaultProjectBuilder.build (DefaultProjectBuilder.java:392)
    at org.apache.maven.project.collector.DefaultProjectsSelector.selectProjects (DefaultProjectsSelector.java:61)
    at org.apache.maven.project.collector.RequestPomCollectionStrategy.collectProjects (RequestPomCollectionStrategy.java:49)
    at org.apache.maven.graph.DefaultGraphBuilder.getProjectsForMavenReactor (DefaultGraphBuilder.java:361)
    at org.apache.maven.graph.DefaultGraphBuilder.build (DefaultGraphBuilder.java:100)
    at org.apache.maven.DefaultMaven.buildGraph (DefaultMaven.java:632)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:254)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:228)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:153)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:958)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:253)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.apache.maven.it.Embedded3xLauncher.run (Embedded3xLauncher.java:218)
    at org.apache.maven.it.Verifier.executeGoals (Verifier.java:1466)
    at org.apache.maven.it.Verifier.executeGoals (Verifier.java:1374)
    at org.apache.maven.buildcache.its.Issue87Test.simple (Issue87Test.java:54)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke (DirectMethodHandleAccessor.java:103)
    at java.lang.reflect.Method.invoke (Method.java:580)
    at org.junit.platform.commons.util.ReflectionUtils.invokeMethod (ReflectionUtils.java:728)
    at org.junit.jupiter.engine.execution.MethodInvocation.proceed (MethodInvocation.java:60)
    at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed (InvocationInterceptorChain.java:131)
    at org.junit.jupiter.engine.extension.TimeoutExtension.intercept (TimeoutExtension.java:156)
    at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod (TimeoutExtension.java:147)
    at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod (TimeoutExtension.java:86)
    at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0 (InterceptingExecutableInvoker.java:103)
    at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0 (InterceptingExecutableInvoker.java:93)
    at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed (InvocationInterceptorChain.java:106)
    at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed (InvocationInterceptorChain.java:64)
    at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke (InvocationInterceptorChain.java:45)
    at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke (InvocationInterceptorChain.java:37)
    at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke (InterceptingExecutableInvoker.java:92)
    at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke (InterceptingExecutableInvoker.java:86)
    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7 (TestMethodTestDescriptor.java:218)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod (TestMethodTestDescriptor.java:214)
    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute (TestMethodTestDescriptor.java:139)
    at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute (TestMethodTestDescriptor.java:69)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6 (NodeTestTask.java:151)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8 (NodeTestTask.java:141)
    at org.junit.platform.engine.support.hierarchical.Node.around (Node.java:137)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9 (NodeTestTask.java:139)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively (NodeTestTask.java:138)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute (NodeTestTask.java:95)
    at java.util.ArrayList.forEach (ArrayList.java:1596)
    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll (SameThreadHierarchicalTestExecutorService.java:41)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6 (NodeTestTask.java:155)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8 (NodeTestTask.java:141)
    at org.junit.platform.engine.support.hierarchical.Node.around (Node.java:137)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9 (NodeTestTask.java:139)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively (NodeTestTask.java:138)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute (NodeTestTask.java:95)
    at java.util.ArrayList.forEach (ArrayList.java:1596)
    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll (SameThreadHierarchicalTestExecutorService.java:41)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6 (NodeTestTask.java:155)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8 (NodeTestTask.java:141)
    at org.junit.platform.engine.support.hierarchical.Node.around (Node.java:137)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9 (NodeTestTask.java:139)
    at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute (ThrowableCollector.java:73)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively (NodeTestTask.java:138)
    at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute (NodeTestTask.java:95)
    at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit (SameThreadHierarchicalTestExecutorService.java:35)
    at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute (HierarchicalTestExecutor.java:57)
    at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute (HierarchicalTestEngine.java:54)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute (EngineExecutionOrchestrator.java:198)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute (EngineExecutionOrchestrator.java:169)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute (EngineExecutionOrchestrator.java:93)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0 (EngineExecutionOrchestrator.java:58)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams (EngineExecutionOrchestrator.java:141)
    at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute (EngineExecutionOrchestrator.java:57)
    at org.junit.platform.launcher.core.DefaultLauncher.execute (DefaultLauncher.java:103)
    at org.junit.platform.launcher.core.DefaultLauncher.execute (DefaultLauncher.java:85)
    at org.junit.platform.launcher.core.DelegatingLauncher.execute (DelegatingLauncher.java:47)
    at org.apache.maven.surefire.junitplatform.LazyLauncher.execute (LazyLauncher.java:56)
    at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute (JUnitPlatformProvider.java:184)
    at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests (JUnitPlatformProvider.java:148)
    at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke (JUnitPlatformProvider.java:122)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess (ForkedBooter.java:385)
    at org.apache.maven.surefire.booter.ForkedBooter.execute (ForkedBooter.java:162)
    at org.apache.maven.surefire.booter.ForkedBooter.run (ForkedBooter.java:507)
    at org.apache.maven.surefire.booter.ForkedBooter.main (ForkedBooter.java:495)
[ERROR]   
[ERROR]   The project org.apache.maven.caching.test.mbuildcache-87:module3:0.0.1-SNAPSHOT (/home/rhousni/projects/maven-build-cache-extension/target/mvn-cache-tests/Issue87Test/simple/project/top/module3/pom.xml) has 1 error
[ERROR]     'build.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-dependency-plugin @ line 45, column 13
[ERROR] 
[ERROR] Re-run Maven using the '-X' switch to enable verbose output
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
```

Reply:
@AlexanderAshitkin I think this validation has been introduced with Maven 4. Should we make a separate integration test running on Maven 3 only? Are tests running on multiple maven versions?